File: LeetCode_Algorithm/242/242.py
from ast import Try


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # sol1 O(nlogn), O(n) 67~106ms 56~10%
        # if len(s) != len(t): return False
        # s_sort = sorted(list(s))
        # t_sort = sorted(list(t))

        # for i in range(len(s_sort)):
        #     if s_sort[i] != t_sort[i]:
        #         return False
        # return True

        # sol2 O O(n), O(1) 68~86ms 54~28%

        if len(s)!= len(t): return False
        hashMap_s = {}
        hashMap_t = {}
        for c in s:
            if c not in hashMap_s:
                hashMap_s[c] = 1 
            else:
                hashMap_s[c] += 1

        for c in t:
            if c not in hashMap_t:
                hashMap_t[c] = 1 
            else:
                hashMap_t[c] += 1       

        for k in hashMap_s:
            if k not in hashMap_t:
                return False
            if hashMap_s[k] != hashMap_t[k]:
                return False

        return True
        


                

print(Solution().isAnagram("rat", "cat"))

File: LeetCode_Algorithm/242/242.js
var isAnagram = function(s, t) {
  // sol1 O(nlogn), O(n) 96~124ms 75~43%
  // if (s.length != t.length) { return false }
  // sArray = s.toLowerCase().split("").sort()
  // tArray = t.toLowerCase().split("").sort()

  // for (let i=0; i < sArray.length; i++){
  //   if (sArray[i] != tArray[i]) { return false }
  // }
  // return true

  // sol2 O(n), O(n) 94~116ms 76~54%
  if (s.length != t.length) { return false }
  sHashMap = {}
  tHashMap = {}

  for (let c of s){
    if (sHashMap[c] == undefined) {
      sHashMap[c] = 1
    }else{
      sHashMap[c] += 1
    }
  }

  for (let c of t){
    if (tHashMap[c] == undefined){
      tHashMap[c] = 1
    }else{
      tHashMap[c] += 1
    }
  }

  for (let k in sHashMap){
    if (sHashMap[k] != tHashMap[k]) { return false }
  }
  return true


};

console.log(isAnagram("cbt", "cat"))